crates/ecs/src/lib.rs

```diff
@@ -220,38 +221,89 @@ impl Application for BasicApplication {
-            .map_err(BasicApplicationError::Execution)
+        let runner = self.into_tickable::<E, S>()?;
+        while let Err(TryRecvError::Empty) = shutdown_receiver.try_recv() {
+            runner.tick()?;
+        }
+        Ok(())
     }
 }
 
-/// A way of executing a `ecs::Schedule`.
-pub trait Executor<'systems>: Default {
-    /// Executes systems in a world according to a given schedule,
-    /// until the `Sender<()>` corresponding to the given `shutdown_receiver` is dropped.
-    fn execute<S: Schedule<'systems>>(
-        &mut self,
-        schedule: S,
-        world: &'systems World,
-        shutdown_receiver: Receiver<()>,
-    ) -> ExecutionResult<()>;
+/// Something that can be turned into being able to run one tick at a time.
+pub trait IntoTickable {
+    /// The type of errors returned the object.
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// The runnable version of the object.
+    type Runnable<E: Executor, S: Schedule>: Tickable<E, S>;
+
+    /// Prepares for execution
+    fn into_tickable<E: Executor, S: Schedule>(self) -> Result<Self::Runnable<E, S>, Self::Error>;
+}
+
+/// Something able to run one tick at a time.
+pub trait Tickable<E: Executor, S: Schedule> {
+    /// The type of errors returned the object.
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Runs a single tick, i.e. a single iteration.
+    fn tick(&self) -> Result<(), Self::Error>;
+}
 
+/// A way of executing a `ecs::Schedule`.
+pub trait Executor: Default {
     /// Executes systems in a world according to a given schedule for a single tick,
     /// meaning all systems get to execute once.
-    fn execute_once<S: Schedule<'systems>>(
+    fn execute_once<S: Schedule>(
         &mut self,
         schedule: &mut S,
-        world: &'systems World,
+        world: &Arc<World>,
     ) -> ExecutionResult<()>;
 }
 
+/// A way of running applications.
+#[derive(Debug)]
+pub struct ApplicationRunner<E: Executor, S: Schedule> {
+    world: Arc<World>,
+    executor: RefCell<E>,
+    schedule: RefCell<S>,
+}
+
+impl IntoTickable for BasicApplication {
+    type Error = BasicApplicationError;
+    type Runnable<E: Executor, S: Schedule> = ApplicationRunner<E, S>;
+
+    fn into_tickable<E: Executor, S: Schedule>(self) -> Result<Self::Runnable<E, S>, Self::Error> {
+        let executor = E::default();
+        let schedule = S::generate(self.systems).map_err(ScheduleGeneration)?;
+        Ok(ApplicationRunner {
+            world: Arc::new(self.world),
+            executor: RefCell::new(executor),
+            schedule: RefCell::new(schedule),
+        })
+    }
+}
+
+impl<E, S> Tickable<E, S> for ApplicationRunner<E, S>
+where
+    S: Schedule,
+    E: Executor,
+{
+    type Error = BasicApplicationError;
+
+    fn tick(&self) -> Result<(), Self::Error> {
+        let mut executor = self.executor.borrow_mut();
+        let mut schedule = self.schedule.borrow_mut();
```

Comment:
Question: 
Does the schedule need to be mutable?

Reply:
Yes, because the executors need the schedule to update its state in order for it to keep track of which systems have finished executing. See here:
https://github.com/martinjonsson01/recs/blob/f406ecdb6fc50b418bf0e3174c19dd2e98c79b14/crates/scheduler/src/executor/mod.rs#L192-L200
on line 200, a call to a function on `schedule` requires `&mut self`.